### PR TITLE
Feat/rlp fsm

### PR DESCRIPTION
**Pull request by kunxian-xia**
### Description

This PR is meant to implement a new RLP circuit which make it easier to support new tx type.
The design doc is located at [here](https://hackmd.io/@kunxian-xia/HJzPo54ds).

### Issue Link

[_link issue here_]

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### Contents

- [x] rlp circuit constraints
- [x] rlp circuit witness gen

### Rationale

[_design decisions and extended information_]

### How Has This Been Tested?

<hr>

### CODE REVIEW

File: zkevm-circuits/src/rlp_circuit_fsm.rs
Comment (kunxian-xia) [inline zkevm-circuits/src/rlp_circuit_fsm.rs:233]: This might not be true. Conside the corner case in which `tx_id > 0 && tx_id' = 0`, the boolean `is_not_padding = true`, but clearly `tx_id != tx_id + 1`.

File: zkevm-circuits/src/rlp_circuit_fsm.rs
Comment (kunxian-xia) [inline zkevm-circuits/src/rlp_circuit_fsm.rs:494]: This condition's degree is larger than 9.

File: zkevm-circuits/src/rlp_circuit_fsm.rs
Comment (roynalnaruto) [reply zkevm-circuits/src/rlp_circuit_fsm.rs:53]: Yes, I wrote the comment before fully understanding the `tag` and `tag_next` layout. Will update 👍

File: zkevm-circuits/src/rlp_circuit_fsm.rs
Comment (roynalnaruto) [reply zkevm-circuits/src/rlp_circuit_fsm.rs:233]: Also, just noticed that the constraint itself is incorrect (there is no `+ 1.expr()`).

I know it won't satisfy the edge case. I was expecting a failure here, and then adding another condition to it, i.e. `next not padding` and a separate conditional check for `next is padding`

File: bus-mapping/src/circuit_input_builder/transaction.rs
Comment (noel2004) [inline bus-mapping/src/circuit_input_builder/transaction.rs:372]: `rlp_unsigned_bytes` is the bytes used for signing, and it is also used for calculating l1fee. However, l1 fee cost is base on the pre-EIP155 rlp bytes so the l1 fee obtained from this field is not consistent with l2geth.

Comment (kunxian-xia) [reply bus-mapping/src/circuit_input_builder/transaction.rs:372]: check the discussion in this issue #515.

File: testool/src/statetest/executor.rs
Comment (roynalnaruto) [inline testool/src/statetest/executor.rs]: This should be:
```suggestion
                rlp_bytes: rlp_signed,
                rlp_unsigned_bytes: rlp_unsigned,
```
?

File: zkevm-circuits/src/rlp_circuit_fsm/util.rs
Comment (roynalnaruto) [inline zkevm-circuits/src/rlp_circuit_fsm/util.rs:1]: Why do we need a separate `IsZeroConfig`?

File: zkevm-circuits/src/witness/tx.rs
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/tx.rs]: nitpick: might be better to replace with a reason such as
```suggestion
                    tx_type => unreachable!("tx type: {:?} not supported", tx_type),
```

File: zkevm-circuits/src/witness/tx.rs
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/tx.rs]: What do you mean by this? I didn't understand..

File: zkevm-circuits/src/witness/tx.rs
Comment (roynalnaruto) [inline zkevm-circuits/src/witness/tx.rs:401]: It might be a good idea to refactor this loop in a later follow-up PR, especially the `match cur.state` part, so that each "current state" is handled by a function, just to make reviewing easier. As of now, I don't really see any mistakes in the witness generation logic though :)

Comment (roynalnaruto) [reply zkevm-circuits/src/witness/tx.rs:401]: Also, in general I would prefer replacing `if..else` with `match` so that its more readable. But that's a minor point for now.

**Review by roynalnaruto — APPROVED** [review]
Looks good to me, although I feel the witness generation might be a bit too complicated and can get messy in that single `loop`. If possible, there should be a more readable way to implementing it which can be done in a follow-up PR.

File: zkevm-circuits/src/rlp_circuit_fsm/util.rs
Comment (kunxian-xia) [reply zkevm-circuits/src/rlp_circuit_fsm/util.rs:1]: Because the `IsZeroChip` provided by the gadgets module can only give us `is_zero_expression` at current row. In the case of RLP, it will be desireable to access the `is_zero_expression` at any rotation.

File: testool/src/statetest/executor.rs
Comment (kunxian-xia) [reply testool/src/statetest/executor.rs]: fixed in https://github.com/scroll-tech/zkevm-circuits/pull/488/commits/f3f656e20de67c5028dbb1ce1ea0ced30b35c6b2

File: zkevm-circuits/src/witness/tx.rs
Comment (kunxian-xia) [reply zkevm-circuits/src/witness/tx.rs]: fixed in https://github.com/scroll-tech/zkevm-circuits/pull/488/commits/f3f656e20de67c5028dbb1ce1ea0ced30b35c6b2

File: zkevm-circuits/src/witness/tx.rs
Comment (kunxian-xia) [reply zkevm-circuits/src/witness/tx.rs]: refined in the https://github.com/scroll-tech/zkevm-circuits/pull/488/commits/f3f656e20de67c5028dbb1ce1ea0ced30b35c6b2. The reason that cur_rom_row is of type `vector` is when we are decoding vectors we cannot decide the next tag is `EndVector` or not. After we parsed the current tag, if the remaining bytes in this layer is zero, then we know the next tag is `EndVector`.

**Comment by kunxian-xia** [comment]
This PR cannot be merged before we make the final decision on how we impl the `rlpEncode(tx)` function. For more details, please refer to https://github.com/scroll-tech/zkevm-circuits/issues/515#issuecomment-1566513645.

File: zkevm-circuits/src/witness/tx.rs
Comment (kunxian-xia) [inline zkevm-circuits/src/witness/tx.rs]: We should use another mocked evm word for testing. For example, changing this to `0x101` will fail the unit test.

**Comment by kunxian-xia** [comment]
@noel2004 Can you update the `tx_l1_fee` gadgets according to the change of `rlpEncode`?

**Comment by noel2004** [comment]
> @noel2004 Can you update the `tx_l1_fee` gadgets according to the change of `rlpEncode`?

ok. I think it would better to separate it in another PR?

**Comment by kunxian-xia** [comment]
Yeah, I agree with that. I have turned off the `tx_l1_fee` unit tests in evm circuit (as they are failing after I modify how we calculated the `tx_data_gas_cost`)

**Comment by kunxian-xia** [comment]
@lispc This PR is ready for merge.

File: zkevm-circuits/src/witness/block.rs
Comment (lispc) [inline zkevm-circuits/src/witness/block.rs]: need to fix?

Comment (kunxian-xia) [reply zkevm-circuits/src/witness/block.rs]: I will fix it in the tx circuit PR as this const var `TX_LEN` is defined in `tx_circuit.rs` and I have comment the tx_circuit out in the `lib.rs`

**Comment by lispc** [comment]
disabling tx circuit in super circuit seems not good.  I would prefer either of the following 2 solutions

(1) wait for tx circuit, then merge this new rlp and tx in one single PR
(2) adjust this PR so it contains only new rlp implementation but will not remove old codes. Super circuit remain unchanged, namely new rlp circuit is only added but not "enabled".

**Comment by kunxian-xia** [comment]
Ok. I will push the updates to tx_circuit in this PR later and re-enable the tx_circuit.

**Comment by kunxian-xia** [comment]
This pr is reopened at #528 because I force updated the `feat/rlp-fsm` branch and the CI does not run again.

**Pull request closed without merging**
